Terraform Registry publishing FAQ (#223)
hashicorp/terraform-docs-common: website/docs/registry/faq.mdx

---
page_title: Terraform Registry FAQ
description: Troubleshooting common Terraform Registry problems
---

## Common publishing issues

### How can I fix a formatting issue in the documentation for my provider?

If your provider's documentation is not showing up in the registry the way you would like it to, you can use the [documentation preview tool](https://registry.terraform.io/tools/doc-preview) to see how the Terraform Registry will format your documentation _before_ you commit the documentation to a new provider release. [More information about how to build provider documentation is available here](/terraform/registry/providers/docs).

> If you have already released a version of your provider and you are unhappy with the appearance of it's documentation, [we recommend you fix the documentation issue and include it in a new release](#how-can-I-correct-a-bug-typo-or-documentation-error-in-a-version-of-my-provider-module-or-policy-library-once-it-has-been-published).

Reviewer: Should we explicitly say

"You cannot update your documentation without a new release." "we recommend" seems a bit like there are other options

If any users have already begun using a version of your provider, module, or policy library, removing the version from the Terraform Registry will break their Terraform configuration. They may be unable to work with their infrastructure until they can find a suitable replacement version of your integration. For this reason, **we recommend that you never remove previously published versions of providers, modules, or policy libraries**.

There may be situations where it is important that a version be removed from the Registry. Perhaps a version of a provider has a critical security vulnerability, or you may have extremely high confidence that no one is using a version of your module that doesn't work as you intended it to.

In these situations, your options for removing previously published versions vary depending on the type of integration:

- **Modules**: The owner of a module can use the Terraform Registry website to remove a specific version, all versions of a specific module provider, or remove the module entirely. [More details on this workflow can be found here](/terraform/registry/modules/publish#removing-published-content).

> - **Providers**: There is no manual process for removing a version of a provider. [Reach out to us via email](/terraform/registry#getting-help) if it is critical that a version be deleted.

Author: Yep, I wanted to link to that location so folks would read through these disclaimers.

- **Policy Libraries**: The owner of a policy library can completely remove it from the Registry, through the website. [More details on this workflow can be found here](/terraform/registry/policy-libraries/publishing#removing-a-policy-library).

### How can I correct a bug, typo, or documentation error in a version of my provider, module, or policy library once it has been published?

> If there is a problem with a certain version, the best resolution is to publish a new release with the issue fixed, incrementing the patch version (e.g. from `1.2.0` to `1.2.1`).

Reviewer: "...according to semantic versioning practices (link)"

### How do I change the GPG key that I use to sign my provider?

The public GPG keys that you upload to the registry serve two purposes:

1. When you publish a new version of a provider, the Registry can use this GPG key to verify the signature file you created. This lets the Registry confirm that the binary files were, in fact, compiled by you.
2. When someone download a version of your provider from the Registry, they can verify that the binary files that they download have not been tampered with since you built them and generated the signature file.

If you have previously used a certain GPG key to sign and publish providers, and you now wish to use a new one, you need only upload the new public GPG key, and use its private key to sign provider releases from now on.

You do not need to remove any old GPG keys. Leaving the old key in place allows consumers of old versions of your provider to continue using those versions.

### How do I remove a public GPG key from the Registry that has been used for publishing a provider?

Generally, you should not ever remove a public GPG key from the Registry. Because of this, we don't provide a way for you to do it yourself.

> If you were to remove a public GPG key that has been used to sign previous versions of a provider, consumers would be unable to use those older versions, because they would be unable to verify the signature files that you built alongside the release.

Reviewer: signature files that you "published" ?

Author: Yeah the fact is that the check is against a signature "file" that's part of the release, but the wording isn't great. I went with:

> If you were to remove a public GPG key that has been used to sign previous versions of a provider, consumers would be unable to use those older versions, **because they would be unable to verify the authenticity of the binary file they downloaded.**

## Can I delete a GitHub tag or release for a published version?

We _do not recommend_ removing content from a GitHub repository you previously published to the Terraform Registry.

The Terraform Registry effectively acts as an index for Terraform-related content hosted on GitHub. When you make a new tag or release in the GitHub repository that holds your Terraform integration, the Registry automatically creates an index record for that content. Think of an index record as a pointer to those files in your GitHub repository.

The Terraform Registry is unaware if you remove a file, a tag, a release, or other content from GitHub. The Registry could still host index records pointing to now-removed content. As a result, someone using your provider or module may have their Terraform configuration break because the Registry is referencing GitHub files that are no longer available.

If you find a typo, bug, or other issues with a published version of your provider, module, or policy library, [we recommend fixing the problem and publishing a new release](#how-do-i-correct-bugs-in-published-versions).

> ## Can I change the name of my integration?

Reviewer 2: eh maybe "terraform registry artifact" the first time and then just "artifact" thereafter?

> ## Can I change the name of my integration?

Reviewer 2: that's ok :)

For this reason, we strongly recommend _against_ moving or renaming an existing integration. Instead, it's often a better choice to create a new repository, representing the new name of your integration while leaving the old content in place.

For further help, [reach out to us via email](/terraform/registry#getting-help).

## Can I rename the GitHub user or organization that owns my integration's repository?

When you change the name of a GitHub user or organization, GitHub automatically redirects requests for the old user or organization to the new one. This means there is no interruption for your integration's consumers, who can continue using the old name. If you need to change the name of your integration, [see the above answer](#can-i-change-the-name-of-my-integration).

If you have changed your publishing Github user or organization name, we _recommend reclaiming_ any names you previously published from. If someone else claims your old user or organization name, they could build repositories that "capture" any redirected requests from the Registry. At that point, your consumers would be unable to download your integration.

## What does resyncing an integration do?

> Terraform integrations owners have access to a **Resync** button on their [providers](/terraform/registry/providers/publishing#publishing-to-the-registry), [modules](/terraform/registry/modules/publish#releasing-new-versions), and [policy libraries](/terraform/registry/policy-libraries/publishing#releasing-policy-library-versions). Clicking **Resync** instructs the Terraform Registry to check an integration's source repository on GitHub, so the Registry:

Reviewer 3: ```suggestion
Owners of Terraform integrations have access to a **Resync** button on their [providers](/terraform/registry/providers/publishing#publishing-to-the-registry), [modules](/terraform/registry/modules/publish#releasing-new-versions), and [policy libraries](/terraform/registry/policy-libraries/publishing#releasing-policy-library-versions). Clicking **Resync** instructs the Terraform Registry to check an integration's source repository on GitHub, so the Registry:
```

---
page_title: Terraform Registry FAQ
description: Troubleshooting common Terraform Registry problems
---

## Common publishing issues

### How can I fix a formatting issue in the documentation for my provider?

If your provider's documentation is not showing up in the registry the way you would like it to, you can use the [documentation preview tool](https://registry.terraform.io/tools/doc-preview) to see how the Terraform Registry will format your documentation _before_ you commit the documentation to a new provider release. [More information about how to build provider documentation is available here](/terraform/registry/providers/docs).

> If you have already released a version of your provider and you are unhappy with the appearance of it's documentation, [you should fix the documentation issue in your repository, and then include it in a new release](#how-can-I-correct-a-bug-typo-or-documentation-error-in-a-version-of-my-provider-module-or-policy-library-once-it-has-been-published).

Reviewer 4: (and we'll need to fix the link if you accept my above header rephrase suggestion!)

### How do I remove a version of my provider, module, or policy library that has already been published?

> If any users have already begun using a version of your provider, module, or policy library, removing the version from the Terraform Registry will break their Terraform configuration. They may be unable to work with their infrastructure until they can find a suitable replacement version of your integration. For this reason, **we recommend that you never remove previously published versions of providers, modules, or policy libraries**.

Reviewer 4: It is a little stylistically out of place, but this does feel super important to emphasize. How would you feel about a compromise between the two (still lots of lovely bold, but not all the words):
```suggestion
If you remove a version of a provider, module, or policy library from the Terraform Registry, you will break the configuration of any individuals using that version. For this reason, **we never recommend removing previously published versions** of providers, modules, or policy libraries.
```

### How can I correct a bug, typo, or documentation error in a version of my provider, module, or policy library once it has been published?

If there is a problem with a certain version, the best resolution is to publish a new release with the issue fixed, incrementing the patch version according to [standard Semantic Versioning practices](https://semver.org/#spec-item-6) (e.g. from `1.2.0` to `1.2.1`).

### Can I remove a version of a provider, module, or policy library by deleting the release or tag from GitHub?

**We recommend you never remove content from a GitHub repository that has already been published to the Terraform Registry**.

The Terraform Registry acts as a kind of index for Terraform-related content that is hosted on GitHub. When you make a new tag or a new release in a GitHub repository that holds your Terraform integration, the Registry automatically creates an index record for a new content. Think of that index record as simply a pointer to the files that exist in your GitHub repository.

> If you remove a file, or a tag, a release, or other content from GitHub, the Registry will not be made aware that you did so. The Registry may still host index records pointing to the now-removed files or content. As a result, someone using your provider or module may see their Terraform configuration break, since through the Registry it expects to find files on GitHub which are no longer available.

Reviewer 4: Rewriting in active voice and removed a few words for brevity:
```suggestion
The Terraform Registry is unaware if you remove a file, a tag, a release, or other content from GitHub. The Registry could still host index records pointing to now-removed content. As a result, someone using your provider or module may have their Terraform configuration break because the Registry is referencing GitHub files that are no longer available.
```

> If you find a typo, bug, or other issue with a published version of your provider, module, or policy library, [we recommend that you fix the issue and publish a new "patch" version of the integration](#how-can-i-correct-a-bug-or-typo-in-a-version-of-my-provider-module-or-policy-library-once-it-has-been-published).

Reviewer 4: Removed words for brevity:
```suggestion
If you find a typo, bug, or other issues with a published version of your provider, module, or policy library, [we recommend fixing the problem and publishing a new release](#how-can-i-correct-a-bug-or-typo-in-a-version-of-my-provider-module-or-policy-library-once-it-has-been-published).
```

Reviewer 4: (and we'll need to fix the link if you accept my above header rephrase suggestion!)

- Look for any discrepancies between the versions of the integration that exist in the GitHub repository and those that exist in the Registry.

If there are new versions available in GitHub that have not been published in the Registry, those new versions will be published.

If there are versions that have been published in the Registry that no longer exist in GitHub, resync behavior diverges depending on the type of integration:

- For modules, versions that are available in the Registry but no longer available in GitHub will be removed from the Registry. This may confuse any consumers who have come to rely on these versions of your module.

- For providers and policy libraries, published versions in the Registry are not removed, even if they no longer exist in GitHub. This state can cause certain versions of the integrations to be unusable. [We recommend that you never delete files, tags, or releases from GitHub once they've been published to the Registry](#can-i-remove-a-version-of-a-provider-module-or-policy-library-by-deleting-the-release-or-tag-from-github).

> Please note that the resync procedure will not remove or update versions of the integration that have already been published.

Author: ```suggestion
```

### I created a new version of my provider, module, or policy library on GitHub and it didn't show up in the Registry. What should I do?

The first thing to try is to [resync your integration with Github](#what-does-it-mean-to-resync-my-provider-module-or-policy-library). While logged in to the Registry, browse to the page for your integration. You can find a button to resync in the "Manage Module" or "Manage Policy Library" dropdown menus for those types of assets, or in the "Settings" tab for a Terraform provider.

If that doesn't work, or if publishing problems occur again _after_ you've done a resync, you should investigate the webhooks that have been established on your GitHub repository. You can locate these webhooks in your repository's settings page:

```
https://github.com/{your github username}/{your repository name}/settings/hooks
```

If there are multiple webhooks listed that point to the Registry's URL (`https://registry.terraform.io`), you should remove them all, and then try to resync your integration again using the above instructions.

> Finally, if you're still running into trouble, please [reach out to us via email](/terraform/registry#getting-help) for support. If you are a customer of HashiCorp, work with your enterprise support contact to escalate the issue.

Reviewer 5: Yes, this sounds good @rclark . Weldone!

### How do I change the GPG key that I use to sign my provider?

The public GPG keys that you upload to the registry serve two purposes:

1. When you publish a new version of a provider, the Registry can use this GPG key to verify the signature file you created. This lets the Registry confirm that the binary files were, in fact, compiled by you.
2. When someone download a version of your provider from the Registry, they can verify that the binary files that they download have not been tampered with since you built them and generated the signature file.

If you have previously used a certain GPG key to sign and publish providers, and you now wish to use a new one, you need only upload the new public GPG key, and use its private key to sign provider releases from now on.

You do not need to remove any old GPG keys. Leaving the old key in place allows consumers of old versions of your provider to continue using those versions.

### How do I remove a public GPG key from the Registry that has been used for publishing a provider?

Generally, you should not ever remove a public GPG key from the Registry. Because of this, we don't provide a way for you to do it yourself.

> If you were to remove a public GPG key that has been used to sign previous versions of a provider, consumers would be unable to use those older versions, because they would be unable to verify the authenticity of the binary file they downloaded.

Reviewer 4: ```suggestion
If you removed a public GPG key that you used to sign previous versions of a provider, consumers would be unable to use those versions because they could not verify the authenticity of the binary file they downloaded.
```

> ### My private GPG key and/or GitHub account was compromised in a security incident. How can I protect my Registry integrations?

Reviewer 4: I kind of like this super short and to the point:
```suggestion
## Compromised GPG keys and GitHub accounts
```

> Please [reach out to us via email](/terraform/registry#getting-help) for support. If you are a customer of HashiCorp, work with your enterprise support contact to escalate the issue.

Reviewer 4: ```suggestion
If you GPG key or Github account has been compromised, please [reach out to us via email](/terraform/registry#getting-help). If you are a customer of HashiCorp, work with your enterprise support contact to escalate the issue.
```

Author: ```suggestion
If your GPG key or Github account has been compromised, please [reach out to us via email](/terraform/registry#getting-help). If you are a customer of HashiCorp, work with your enterprise support contact to escalate the issue.
```

---
page_title: Terraform Registry FAQ
description: Troubleshooting common Terraform Registry problems
---

## Common publishing issues

### How does the Registry identify the latest version of my provider, module, or policy library?

Each version of an integration published in the Terraform Registry has a version identifier, which is a string like `1.2.3`. You create the version identifier when you create a tag (for modules and policy libraries) or a release (for providers) in the GitHub repository where the integration's code lives.

> The Registry requires that these version identifiers are [SemVer compliant](https://semver.org/). Versions are sorted using these identifiers, following [standard SemVer rules for precedence](https://semver.org/#spec-item-11). The version identifier with the greatest precedence is always listed as latest by the Terraform Registry.

Author: The Terraform CLI won't pick a prerelease version. It would make sense to call that out here.

If there is a bug in a certain version, it is usually better to publish another release, incrementing the patch version (i.e. from `1.2.0` to `1.2.1`), rather than to remove the bugged version.

There may be situations where it is important that a version be removed from the Registry. Perhaps a version of a provider has a critical security vulnerability, or you may have extremely high confidence that no one is using a version of your module that doesn't work as you intended it to.

In these situations, your options for removing previously published versions vary depending on the type of integration:

- **Modules**: The owner of a module can use the Terraform Registry website to remove a specific version, all versions of a specific module provider, or remove the module entirely. [More details on this workflow can be found here](/registry/modules/publish#removing-published-content). or [here for policy libraries].

> - **Providers**: There is no manual process for removing a version. Reach out to us via email if it is critical that a version be deleted.

Reviewer: Might want to include that we will take the request under consideration, but it's not guaranteed we will remove it. Not sure if there are GDPR implications to saying something like that though.

- **Policy Libraries**: The owner of a policy library can completely remove it from the Registry, through the website. [More details on this workflow can be found here](/registry/policy-libraries/publishing#removing-a-policy-library).

### Can I remove a version of a provider by deleting the release from GitHub?

**We recommend you never remove content from a GitHub repository that has already been published to the Terraform Registry**.

The Terraform Registry acts as a kind of index for Terraform-related content that is hosted on GitHub. When you make a new tag or a new release in a GitHub repository that holds your Terraform integration, the Registry automatically creates an index record for a new content. Think of that index record as simply a pointer to the files that exist in your GitHub repository.

> If you remove a file, or a tag, a release, or other content from GitHub, the Registry will not be made aware that you did so. The Registry may still host index records pointing to the now-removed files or content. As a result, someone using your provider or module may see their Terraform configuration break, since through the Registry it expects to find files on GitHub which are no longer available.

Reviewer: Would be great to include the recommended path alongside this. Maybe something like:

"Instead, if you discover a flaw in a published version of your module or provider, we recommend publishing a new version with the flaw fixed."

Reviewer: Ah, I see you say this elsewhere in the FAQ - maybe we make it a separate question and then link to it from these other questions?

### What does it mean to "resync" my provider, module, or policy library?

Owners of any of these types of Terraform integrations can find a "resync" button on the webpage for the integrations they own. Triggering a resync instructs the Registry to analyze the integration's source repository on GitHub. During a resync, the registry will:

- Ensure that there is a webhook established on the repository, so that new versions of the integration can be automatically published.
- Create any versions of the integration that exist in the GitHub repository but do not yet exist in the Registry.

> Please note that the resync procedure will not remove or update versions of the integration that have already been published.

Reviewer 5: I dont think this holds for modules. Module versions are mutable, and upserted on every resync. See https://github.com/hashicorp/terraform-registry/blob/947f8e9a8b85cd09a27715e8db5cafb879e2fa88/ingress/github/resync.go#L50-L78

Author: Thanks for pointing this out. The divergent behavior between modules and providers is confusing, and it leads to a problem when in putting together docs like this.

We would probably all agree that the behavior of resync should be predictable across all kinds of registry integrations. So should we:

- Document the divergent behavior first, then change that behavior and the docs later so that it is predictable?
- Document the divergent behavior and then never change it, so that we don't confuse users later?
- Make the behaviors identical first and then write the docs for the predictable behavior that we want to have?

Reviewer: I suggest - document the divergent behaviour, indicate that work is planned to have it converge (in the docs), then update it when the behaviour has converged.

### What does it mean to "resync" my provider, module, or policy library?

Owners of any of these types of Terraform integrations can find a "resync" button on the webpage for the integrations they own. Triggering a resync instructs the Registry to analyze the integration's source repository on GitHub. During a resync, the registry will:

- Make sure that there is a webhook established on the repository, so that new versions of the integration can be automatically published.
- Look for any discrepancies between the versions of the integration that exist in the GitHub repository and those that exist in the Registry.

If there are new versions available in GitHub that have not been published in the Registry, those new versions will be published.

If there are versions that have been published in the Registry that no longer exist in GitHub, resync behavior diverges depending on the type of integration:

- For modules, versions that are available in the Registry but no longer available in GitHub will be removed from the Registry. This may confuse any consumers who have come to rely on these versions of your module.

> 

Reviewer 6: These are some additional edge cases, that have come up as 'publishing' issues for modules. Could we highlight them with a line or two on what is the expected system behavior in such cases in terms of resync.
        Moving a Repository (to a different GH org/namespace)
	Rename module or provider
